### PR TITLE
starting cell made by 1 hex of chyto, nuclues in editor, locked organelles

### PR DESCRIPTION
**Pull request by Dak2896**
- Starting cell is now made by 1 hex of cytoplasm.
- Nucleus can be added with editor by 100 mp.
- Can be added one nucleus.
- Some organelle need nucleus to be added to cell.

TODO:
- GUI should show player that some nucleus are locked and need nucleus.
- Nucleus need 2d icon.
- Nucleus need description.

### CODE REVIEW

**Comment by Untrustedlife** [comment]
This is looking alright, one thing, of the player has no cucleus the game should treat them as bacteria, eg, halve there size.

File: scripts/microbe_stage/microbe_editor/microbe_editor.as [hunk 1]
```diff
@@ -228,6 +230,8 @@ class MicrobeEditor{
 
     private void _addOrganelle(PlacedOrganelle@ organelle)
     {
+        if((!organelle.organelle.needNucleus && organelle.organelle.name != "nucleus") || (organelle.organelle.needNucleus && isNucleusPresent) || (organelle.organelle.name == "nucleus" && !isNucleusPresent))
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe_editor/microbe_editor.as]: You might want to turn this condition around and return early as otherwise you'd need to indent this whole function.

File: scripts/microbe_stage/microbe_editor/microbe_editor.as [hunk 2]
```diff
@@ -274,11 +281,19 @@ class MicrobeEditor{
...
+                //check if nucleus has been removed
+                editor.getNucleusIsPresent();
             });
 
         @action.data["organelle"] = organelle;
 
         enqueueAction(action);
+        }
+        else
+        {   // debug only
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe_editor/microbe_editor.as]: Comment shouldn't be on the same line as '{'

File: scripts/microbe_stage/microbe_editor/microbe_editor.as [hunk 3]
```diff
@@ -563,6 +578,26 @@ class MicrobeEditor{
...
+        for(uint i = 0; i < editedMicrobe.length(); ++i){
+            auto organelle = cast<PlacedOrganelle>(editedMicrobe[i]);
+            auto name = organelle.organelle.name;
+            if (name=="nucleus"){
+                isNucleusPresent = true;
+                LOG_INFO("Nucleus is present!");
+                break;
+            }
+            else {
+                isNucleusPresent = false;
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe_editor/microbe_editor.as]: Doesn't this line completely break what this method is attempting to do.

File: scripts/microbe_stage/microbe_editor/microbe_editor.as [hunk 4]
```diff
@@ -994,6 +1029,8 @@ class MicrobeEditor{
     // private auto nextMicrobeEntity;
     // private dictionary occupiedHexes;
 
+    // determine if nucleus is present in current cell
+    private bool isNucleusPresent = false;
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe_editor/microbe_editor.as]: I don't think this should be a variable as keeping it up to date is a lot of work. Instead `getNucleusIsPresent` should return true if it finds the nucleus. And maybe it could be renamed `checkIsNucleusPresent`

File: scripts/microbe_stage/organelle.as [hunk 5]
```diff
@@ -226,6 +227,9 @@ class Organelle{
 
     //! Cost in mutation points
     int mpCost = 0;
+
+    //! boolean that represent if organelle need presence of nucleus
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/organelle.as]: This should clarify that this is checked by the editor when attempting to place. And a good thing would also be to use this in the future in autoevo.

Edit and a bit more grammatically pleacing name would be `needsNucleus`

File: scripts/microbe_stage/microbe_editor/microbe_editor.as (hunk 3 above)
Comment (Dak2896) [reply scripts/microbe_stage/microbe_editor/microbe_editor.as]: if is not found variable is set to false, if found 1 time is set true and go out with break. of course if the function change and become with return value this set is going to die.

Comment (hhyyrylainen) [reply scripts/microbe_stage/microbe_editor/microbe_editor.as]: Okay, I see the break now. It is still very weird to be setting it to false in the loop rather than setting it to false once.

Comment (Dak2896) [reply scripts/microbe_stage/microbe_editor/microbe_editor.as]: yhe actually i could just set it false before loop.

File: scripts/microbe_stage/microbe_editor/microbe_editor.as (hunk 1 above)
Comment (hhyyrylainen) [reply scripts/microbe_stage/microbe_editor/microbe_editor.as]: You have added an IF statement but you didn't indent the function. It's going to start getting pretty tight regards 95 column limit, so I suggest you write that as
```
if(cant place stuff)
    return;
```

instead of indenting the entire function one more level.

Comment (Dak2896) [reply scripts/microbe_stage/microbe_editor/microbe_editor.as]: Ok use the condition that don't allow me to put things instead of good ones.
that are only two, i have nucleus and i'm trying to put it, or i don't have it and i'm trying put something that need it

**Comment by Dak2896** [comment]
i put at line 236 a blank line, if you think it's not necessary i remove it.

**Comment by Dak2896** [comment]
If this last commit is not useful i just revert it

**Comment by Dak2896** [comment]
My runcodeformatting don't give any errors, i do not understand what's wrong here

**Comment by Untrustedlife** [comment]
You weren't supposed to make that grayscale change yourself you were supposed to pull master,. now we might get merge conflicts /:

File: scripts/gui/thrive_style.css [hunk 6]
```diff
@@ -235,8 +235,6 @@ video {
   flex-direction: column;
 }
 
-
```
Comment (Untrustedlife) [inline scripts/gui/thrive_style.css]: This is the random white space i was talking about.

File: scripts/gui/thrive_style.css [hunk 7]
```diff
@@ -1109,6 +1100,16 @@ video {
...
+#NucleusIcon {  
+  position: relative; 
+  left: calc(50% - 30px); 
+  width: 60px;  
+  height: 60px; 
+  background-size: contain; 
+  background-image: url("../../Textures/gui/bevel/NucleusIcon.png");  
+}
+
+
```
Comment (Untrustedlife) [inline scripts/gui/thrive_style.css]: what was th epoint of moving the class here?

File: scripts/gui/thrive_style.css [hunk 8]
```diff
@@ -1108,6 +1099,15 @@ video {
   background-image: url("../../Textures/gui/bevel/NucleusIcon.png");  
 }
 
+#PlastidIcon {
```
Comment (Untrustedlife) [inline scripts/gui/thrive_style.css]: why did you move this one too???

File: scripts/gui/thrive_style.css (hunk 7 above)
Comment (Untrustedlife) [reply scripts/gui/thrive_style.css]: Actually looks like the plastid being moved didnt register in the diff so you were moving it back to a previous position, so im okay with this.

File: scripts/gui/thrive_gui.html [hunk 9]
```diff
@@ -404,6 +404,14 @@
...
                 <div id="PlastidIcon">
                 </div>Nitrogen Fixing Plastid<br>80 MP</td>
+              <td id="addNucleus" class="OrganelleListItem">
+                <span class="tooltiptext">
+                Eukariote Nucleus<hr><br> Cost: 100 mutation points<hr><br>
+                Performs Process: Eukariote Evolution<hr><br>
+                Allows for synthesis of RNA, MRNA, allow add chloroplasts and mitochondria
+                </span>
+                <div id="NucleusIcon">
+                </div>Eukariote Nucleus <br>100 MP</td>
```
Comment (Untrustedlife) [inline scripts/gui/thrive_gui.html]: This html change looks good,  but Eukaryote is spelled wrong

File: scripts/microbe_stage/organelle_table.as [hunk 10]
```diff
@@ -659,6 +676,7 @@ void setupOrganelles(){
     nitrogenFixationProtien.chanceToCreate = 0;
     nitrogenFixationProtien.prokaryoteChance = 1;
     nitrogenFixationProtien.mpCost = 15;
+    nitrogenFixationProtien.needsNucleus = false;
```
Comment (Untrustedlife) [inline scripts/microbe_stage/organelle_table.as]: One suggestion i have would be to remove this parameter entirely and just check if prokaryote chance is greater then 0. Would accomplish the same thing with less code.

File: scripts/gui/microbe_editor.mjs [hunk 11]
```diff
@@ -310,6 +319,22 @@ function onExitToMenuClickedEditor(){
     }
 }
 
+function updateGuiButtons(isNucleusPresent){
+    if(!isNucleusPresent) {
+        $("#addMitochondrion").addClass("DisabledButton");
```
Comment (hhyyrylainen) [inline scripts/gui/microbe_editor.mjs]: This doesn't need jQuery, it should be used only when it makes something easier. See the other places where the classlist is changed.

File: scripts/microbe_stage/microbe_editor/microbe_editor.as [hunk 12]
```diff
@@ -563,6 +576,28 @@ class MicrobeEditor{
     }
 
 
+    bool checkIsNucleusPresent()
+    {
+        double lengthMicrobe = double(editedMicrobe.length());
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe_editor/microbe_editor.as]: Why is there an unused variable here?

~~Why is this a double???? the length is of type `uint` and not a floating point type, mixing them is not good.~~

File: scripts/microbe_stage/microbe_editor/microbe_editor.as [hunk 13]
```diff
@@ -563,6 +576,28 @@ class MicrobeEditor{
...
+        for(uint i = 0; i < editedMicrobe.length(); ++i){
+            auto organelle = cast<PlacedOrganelle>(editedMicrobe[i]);
+            auto name = organelle.organelle.name;
+            if (name=="nucleus"){
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe_editor/microbe_editor.as]: I think this line should be: `if(organelle.organelle.name == "nucleus")`

File: scripts/microbe_stage/organelle_table.as [hunk 14]
```diff
@@ -104,6 +104,9 @@ class OrganelleParameters{
 
     //! Cost in mutation points
     int mpCost = 0;
+
+    //!  boolean used in editor to know if organelle that player is try to putneed presence or not of nucleus.
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/organelle_table.as]: There are extra spaces here. Also I'd like comments to be started with a capital letter.

File: scripts/gui/microbe_editor.mjs [hunk 15]
```diff
@@ -310,6 +319,22 @@ function onExitToMenuClickedEditor(){
     }
 }
 
+function updateGuiButtons(isNucleusPresent){
```
Comment (hhyyrylainen) [inline scripts/gui/microbe_editor.mjs]: This function needs to also block the player from clicking the buttons. Perhaps the button click code should check if it contains `DisabledButton` class and if it does then ignore the click.

**Comment by Dak2896** [comment]
Line 245 in microbe_editor.mjs is longer than 95 length i think i have to change the function.

File: scripts/gui/microbe_editor.mjs [hunk 16]
```diff
@@ -229,38 +238,37 @@ function setRedo(enabled){
 }
 
 //! Sends organelle selection to the Game
-function onSelectNewOrganelle(organelle){
+function onSelectNewOrganelle(element){
 
-    if(common.isInEngine()){
+    if(common.isInEngine() && !$(element.element).hasClass("DisabledButton")) {
```
Comment (hhyyrylainen) [inline scripts/gui/microbe_editor.mjs]: This should be in the upper function. This is could for example be called from the editor, so this *must* take the organelle name as a parameter.

**Comment by hhyyrylainen** [comment]
You need to merge *master* (from the main repo) as Github is complaining that "
This branch is out-of-date with the base branch".

Also you should try to make bigger commits with more meaningful commit messages.

**Comment by Dak2896** [comment]
I think to keep up to date my fork i have to close pull request, i'm not finding a way to do it.

**Comment by hhyyrylainen** [comment]
You need to switch to master branch, pull it, switch back and then merge master into your branch.

**Comment by Dak2896** [comment]
So i create new branch, i copy in that branch main repo master, then i merge the main repo master with mine master, and then make pull request?

**Comment by hhyyrylainen** [comment]
No. You should look at more git tutorials. But basically this should work:
```
git remote add main https://github.com/Revolutionary-Games/Thrive.git
git pull main master
```
That should update your fork with commits from the main repo.

**Comment by hhyyrylainen** [comment]
The script file is broken again: https://github.com/Revolutionary-Games/Thrive/pull/704/commits/601768f97e7f9dd342f7302c8a6eaed0850d1149 (the permissions changed)

**Comment by Dak2896** [comment]
i don't know why, to make the merge i have to commit that, i try also stash..

File: scripts/microbe_stage/microbe_editor/microbe_editor.as [hunk 17]
```diff
@@ -228,6 +230,11 @@ class MicrobeEditor{
 
     private void _addOrganelle(PlacedOrganelle@ organelle)
     {
+        // 1 - you put nucleus but you already have it
+        // 2 - you put organelle that need nucleus and you don't have it
+        if((organelle.organelle.name == "nucleus" && checkIsNucleusPresent()) || (organelle.organelle.prokaryoteChance > 0 && !checkIsNucleusPresent()))
```
Comment (Untrustedlife) [inline scripts/microbe_stage/microbe_editor/microbe_editor.as]: Thanks for switchinbgto the previously existing parameters. Okay, now all you need to do is  get this if statement seperated into two lines (simply have the first half on one line and the other half on the other by adding a new line after the || bit

File: scripts/microbe_stage/microbe_editor/microbe_editor.as [hunk 18]
```diff
@@ -232,7 +232,8 @@ class MicrobeEditor{
     {
         // 1 - you put nucleus but you already have it
         // 2 - you put organelle that need nucleus and you don't have it
-        if((organelle.organelle.name == "nucleus" && checkIsNucleusPresent()) || (organelle.organelle.prokaryoteChance > 0 && !checkIsNucleusPresent()))
+        if((organelle.organelle.name == "nucleus" && checkIsNucleusPresent()) ||
+            (organelle.organelle.prokaryoteChance > 0 && !checkIsNucleusPresent()))
```
Comment (Untrustedlife) [inline scripts/microbe_stage/microbe_editor/microbe_editor.as]: Nice

**Comment by Dak2896** [comment]
Ok i make a big mess up with merging, now we are when everything was working without merging with master repo, i add the two things that @Untrustedlife  ask me, they work, and then i try to merge again.
